src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2836,8 +2834,31 @@ module Crystal
       false
     end
 
+    def visit(node : TypeRestrict)
+      value = node.value
+      restriction = node.restriction
+
+      value.accept self
+
+      unless context = match_context
```

Comment:
`freeze_type` looks good. But, I think this way doesn't solve `def foo(x : T.class = String)` problem, and fixing this problem needs real call of `restrict` method, so `match_context` is needed.

> The problem with this approach is that the type is checked once here (below). But what happens if the type of value changes later? (rare, but can happen)

Could you show me sample code?